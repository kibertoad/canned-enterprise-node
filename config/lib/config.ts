import {
  Brand,
  ConfigDefinition,
  ConfigDefinitionMap,
  Country,
  DeepPartial,
  Environment,
  FeatureToggleDefinition,
  FeatureToggleEnvironmentMap
} from './configInterface'

export const dbConfig = Object.freeze({
  default: {
    client: 'pg',
    connection: {
      host: '127.0.0.1',
      user: 'your_database_user',
      password: 'your_database_password',
      database: 'myapp_test'
    }
  }
})

function getMandatoryEnvParam(key: string): string {
  const value = process.env[key]
  if (!value) {
    throw new Error(`Please define environment variable ${key}`)
  }

  return value
}

// Default configuration for all environments. Single keys and values can be overridden in the environment specific configuration
export const defaultConfig: DeepPartial<ConfigDefinition> = {}

// Override the default config values here for environments if needed
export const environmentConfig: ConfigDefinitionMap = {
  default: {
    someExternalService: {
      baseUrl: 'http://service.com',
      auth: {
        username: getMandatoryEnvParam('EXTERNAL_SYSTEM_USERNAME'),
        password: getMandatoryEnvParam('EXTERNAL_SYSTEM_PASSWORD')
      }
    },
    aws: {
      accessKeyId: getMandatoryEnvParam('AWS_ACCESS_KEY_ID'),
      secretAccessKey: getMandatoryEnvParam('AWS_SECRET_ACCESS_KEY')
    }
  },
  production: {
    aws: {
      region: {
        [Brand.MainBrand]: {
          [Country.US]: 'eu-west-1',
          [Country.RUSSIA]: 'eu-north-1'
        },
        [Brand.OtherBrand]: {
          [Country.US]: 'eu-north-1',
          [Country.RUSSIA]: 'eu-west-1'
        }
      }
    }
  },
  staging: {
    aws: {
      region: {
        [Country.US]: 'eu-west-1',
        [Country.RUSSIA]: 'eu-north-1'
      }
    }
  }
}

export const defaultFeatureToggles: DeepPartial<FeatureToggleDefinition> = {
  GLOBAL_FEATURE: {
    [Brand.MainBrand]: false,
    [Brand.OtherBrand]: true
  }
}

export const environmentFeatureToggles: FeatureToggleEnvironmentMap = {
  staging: {
    GLOBAL_FEATURE: {
      [Brand.MainBrand]: true,
      [Brand.OtherBrand]: true
    },
    moduleOne: {
      SOME_FEATURE: [Country.US, Country.RUSSIA]
    },
    moduleTwo: {
      ANOTHER_FEATURE: true
    }
  }
}

export const isProductionEnvironment = process.env.APP_ENV === Environment.production
