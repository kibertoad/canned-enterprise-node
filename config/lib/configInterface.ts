export enum Brand {
  MainBrand = 'mainbrand',
  OtherBrand = 'otherbrand'
}

export enum Country {
  US = 'US',
  RUSSIA = 'RU'
}

export interface Config {
  aws: {
    accessKeyId: string
    secretAccessKey: string
    region: string
  }
  someExternalService: {
    baseUrl: string
    auth: {
      username: string
      password: string
    }
    callTimeout: number
  }
}

export interface ConfigDefinition {
  aws: {
    accessKeyId: StringParam
    secretAccessKey: StringParam
    region: StringParam
  }
  someExternalService: {
    baseUrl: StringParam
    auth: {
      username: StringParam
      password: StringParam
    }
    callTimeout: NumberParam
  }
}

export type ConfigDefinitionMap = DeepPartial<
  { [key in Environment | 'default']: ConfigDefinition }
>

type StringParam =
  | string
  | { [key in Brand]: string | { [key in Country]: string } }
  | { [key in Country]: string }
type BooleanParam = boolean | { [key in Brand]: boolean | { [key in Country]: boolean } | string[] }
type NumberParam = number | { [key in Brand]: number | { [key in Country]: number } | number[] }

export enum Environment {
  production = 'production',
  staging = 'staging',
  test = 'test',
  development = 'development'
}

export interface FeatureToggles {
  moduleOne: {
    SOME_FEATURE: boolean
  }
  moduleTwo: {
    ANOTHER_FEATURE: boolean
  }
  GLOBAL_FEATURE: boolean
}

export interface FeatureToggleDefinition {
  moduleOne: {
    SOME_FEATURE: BooleanParam | string[]
  }
  moduleTwo: {
    ANOTHER_FEATURE: BooleanParam | string[]
  }
  GLOBAL_FEATURE: BooleanParam | string[]
}

export type FeatureToggleEnvironmentMap = DeepPartial<
  { [key in Environment]: FeatureToggleDefinition }
>

export type DeepPartial<T> = {
  [P in keyof T]?: T[P] extends Array<infer U>
    ? Array<DeepPartial<U>>
    : T[P] extends ReadonlyArray<infer U>
    ? ReadonlyArray<DeepPartial<U>>
    : DeepPartial<T[P]>
}
