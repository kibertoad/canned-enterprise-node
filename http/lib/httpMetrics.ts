import { Counter, Histogram } from 'prom-client'
import * as Prometheus from 'prom-client'

let areMetricsInitted = false
export let httpRequestCounter: Counter
export let httpRequestDurationSeconds: Histogram

export function initOutgoingMetrics() {
  if (!areMetricsInitted) {
    httpRequestDurationSeconds = new Prometheus.Histogram({
      name: 'outgoing_http_request_duration_seconds',
      help: 'Duration of HTTP outgoing requests in seconds',
      labelNames: ['request', 'locale', 'service']
    })

    httpRequestCounter = new Prometheus.Counter({
      name: 'outgoing_http_request_count',
      help: 'Outgoing request status code',
      labelNames: ['request', 'statusCode', 'locale', 'service']
    })

    areMetricsInitted = true
  }
}
