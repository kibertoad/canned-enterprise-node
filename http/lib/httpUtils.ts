const IS_DEVELOPMENT = process.env.NODE_ENV === 'development'

export function roundProperties(obj: any): any {
  if (typeof obj !== 'object') {
    return obj
  }
  return Object.keys(obj).reduce((acc, curr) => {
    const currentObject = obj[curr]
    acc[curr] = typeof currentObject === 'number' ? Math.round(obj[curr]) : currentObject
    return acc
  }, {} as Record<string, any>)
}

export function stripCredentialsFromUrl(url: string): string {
  if (typeof url !== 'string' || IS_DEVELOPMENT) {
    return url
  } else {
    return url.replace(/^((?:\w+:)?\/\/)(?:[^@/]+@)/, '$1')
  }
}
