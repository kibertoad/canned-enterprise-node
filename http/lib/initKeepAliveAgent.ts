import { default as HttpAgent, HttpOptions, HttpsAgent, HttpsOptions } from 'agentkeepalive'
import { Agent } from 'http'

const IS_DEVELOPMENT = process.env.NODE_ENV === 'development'

export function initKeepAliveAgent(timeout: number): (url: string) => Agent {
  const options: HttpOptions | HttpsOptions = {
    keepAlive: true,
    freeSocketTimeout: timeout,
    rejectUnauthorized: !IS_DEVELOPMENT
  }
  const httpKeepAliveAgent = new HttpAgent(options)
  const httpsKeepAliveAgent = new HttpsAgent(options)

  return (url: string) => {
    if (url.startsWith('https')) {
      return httpsKeepAliveAgent
    } else {
      return httpKeepAliveAgent
    }
  }
}
