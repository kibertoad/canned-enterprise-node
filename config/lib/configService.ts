import deepMerge from 'deepmerge'
import memoizee from 'memoizee'
import * as dotenv from 'dotenv'

dotenv.config() // This needs to be called before we import config

import {
  environmentConfig,
  defaultConfig,
  environmentFeatureToggles,
  defaultFeatureToggles,
  dbConfig
} from './config'
import { Brand, Config, Environment, FeatureToggles } from './configInterface'
import {
  buildConfigFromDefinition,
  buildFeatureTogglesFromDefinition
} from './internal/configBuilder'

const environment: Environment = (process.env.APP_ENV as Environment) || Environment.development

// ToDo Implement env-specific config
export function getDbConfig() {
  return dbConfig.default
}

export const getConfig = memoizee(getConfigInternal, {
  normalizer: (args: any) => {
    return `${args[0]}-${args[1]}`
  }
})

function getConfigInternal(country: string, brand: Brand): Config {
  if (!Object.values(Environment).includes(environment)) {
    throw new Error(`Unknown environment: ${environment}`)
  }
  const builtDefaultConfig = buildConfigFromDefinition(defaultConfig, country, brand)

  const environmentDefinition = environmentConfig[environment]
  if (!environmentDefinition) {
    return builtDefaultConfig
  }

  const builtConfig = buildConfigFromDefinition(environmentDefinition, country, brand)

  return deepMerge(builtDefaultConfig, builtConfig)
}

export const getFeatures = memoizee(getFeaturesInternal, {
  normalizer: (args: any) => {
    return `${args[0]}-${args[1]}`
  }
})

function getFeaturesInternal(country: string, brand: Brand): FeatureToggles {
  if (!Object.values(Environment).includes(environment)) {
    throw new Error(`Unknown environment: ${environment}`)
  }

  const builtDefaultConfig = buildFeatureTogglesFromDefinition(
    defaultFeatureToggles,
    country,
    brand
  )

  const environmentDefinition = environmentFeatureToggles[environment]
  if (!environmentDefinition) {
    return builtDefaultConfig
  }

  const builtConfig = buildFeatureTogglesFromDefinition(environmentDefinition, country, brand)

  return deepMerge(builtDefaultConfig, builtConfig)
}
