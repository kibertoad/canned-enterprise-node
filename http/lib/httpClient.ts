import { GotOptions, HTTPError, ResponseType } from 'got'
import * as got from 'got'
import defaultGot from 'got'
import { OptionsOfDefaultResponseBody } from 'got/dist/source/create'
import * as http from 'http'
import * as https from 'https'
import { labelValues } from 'prom-client'
import { httpRequestCounter, httpRequestDurationSeconds } from './httpMetrics'
import { roundProperties, stripCredentialsFromUrl } from './httpUtils'

import { initKeepAliveAgent } from './initKeepAliveAgent'
// The timeout value is an educated guess because there is no correct answer.
// Some requests probably can use a higher value.
const DEFAULT_FREE_SOCKET_TIMEOUT = 4500
const getDefaultKeepAliveAgent = initKeepAliveAgent(DEFAULT_FREE_SOCKET_TIMEOUT)

const defaultHeaders = {
  'user-agent': 'MicroserviceX/1.0'
}

const defaultOptions: Partial<OptionsOfDefaultResponseBody> = {
  encoding: 'utf8'
}

export type HttpOptions = {
  requestId?: string
  locale?: string
  requestName?: string
  serviceName?: string

  responseType?: ResponseType
  retry?: number
  method?: 'GET' | 'POST' | 'PUT' | 'PATCH' | 'DELETE'
  agent?: http.Agent | https.Agent
  baseUrl?: string
  qs?: Record<string, any>
  headers?: http.IncomingHttpHeaders
}

type HttpOptionsInternal = {
  json?: Record<string, any> | Record<string, any>[]
  url: string
} & HttpOptions

export declare type ResponseBody = Buffer | string | object

export class HttpClient {
  private readonly got: got.Got

  constructor(gotInstance: got.Got = defaultGot) {
    this.got = gotInstance
  }

  async get<T extends ResponseBody>(
    url: string,
    options: HttpOptions = {}
  ): Promise<got.Response<T>> {
    const response: got.Response<T> = await execute(this.got, {
      url,
      responseType: 'json',
      ...options,
      method: 'GET'
    })
    return response
  }

  async post<T extends ResponseBody>(
    url: string,
    body: Record<string, any> | Record<string, any>[],
    options: HttpOptions = {}
  ): Promise<got.Response<T>> {
    const response: got.Response<T> = await execute(this.got, {
      url,
      responseType: 'json',
      ...options,
      json: body,
      method: 'POST'
    })
    return response
  }

  async patch<T extends ResponseBody>(
    url: string,
    body: Record<string, any>,
    options: HttpOptions = {}
  ): Promise<got.Response<T>> {
    const response: got.Response<T> = await execute(this.got, {
      url,
      responseType: 'json',
      ...options,
      json: body,
      method: 'PATCH'
    })
    return response
  }

  async put<T extends ResponseBody>(
    url: string,
    body: Record<string, any>,
    options: HttpOptions = {}
  ): Promise<got.Response<T>> {
    const response: got.Response<T> = await execute(this.got, {
      url,
      responseType: 'json',
      ...options,
      json: body,
      method: 'PUT'
    })
    return response
  }

  async delete<T extends ResponseBody>(
    url: string,
    options: HttpOptions = {}
  ): Promise<got.Response<T>> {
    const response: got.Response<T> = await execute(this.got, {
      url,
      responseType: 'json',
      ...options,
      method: 'DELETE'
    })
    return response
  }
}

function initHeaders(requestOptions: Partial<HttpOptions> = {}): http.IncomingHttpHeaders {
  const requestHeaders = requestOptions.requestId
    ? { 'X-Request-ID': requestOptions.requestId }
    : {}

  return {
    ...defaultHeaders,
    ...requestHeaders
  }
}

declare type ResolvedOptions = {
  searchParams?: Record<string, any>
} & GotOptions

function mapOptions(options: Partial<HttpOptionsInternal>): ResolvedOptions {
  const resolvedOptions: ResolvedOptions = {
    retry: options.retry ?? 0,
    prefixUrl: options.baseUrl,
    responseType: options.responseType,
    method: options.method
  }

  if (options.qs) {
    resolvedOptions.searchParams = options.qs
  }

  if (options.json) {
    resolvedOptions.json = options.json
  }
  return resolvedOptions
}

function initOptions(options: HttpOptionsInternal): GotOptions {
  const agent = 'agent' in options ? options.agent : getDefaultKeepAliveAgent(options.url)

  const resolvedOptions: GotOptions = mapOptions(options)
  const headers = initHeaders(options)

  console.debug('HTTP request', {
    requestId: options.requestId,
    method: options.method,
    headers,
    url: stripCredentialsFromUrl(options.url)
  })

  return {
    ...defaultOptions,
    ...resolvedOptions,
    agent,
    headers
  }
}

async function execute<T extends ResponseBody>(
  gotInstance: got.Got,
  options: HttpOptionsInternal
): Promise<got.Response<T>> {
  const metricLabels: labelValues = {
    locale: options.locale!,
    ...(options.requestName && { request: options.requestName }),
    ...(options.serviceName && { service: options.serviceName })
  }
  const trackMetrics = !!(options.requestName || options.serviceName)
  const resolvedOptions = initOptions(options)

  let endTimer: (labels?: labelValues) => void
  let start: Date
  if (trackMetrics) {
    start = new Date()
    endTimer = httpRequestDurationSeconds.startTimer()
  }

  try {
    const response = await gotInstance(options.url, resolvedOptions as any)
    if (trackMetrics) {
      endTimer!({ ...metricLabels })
    }
    let duration
    if (trackMetrics) {
      duration = new Date().getTime() - start!.getTime()
      httpRequestCounter.inc(
        {
          statusCode: response.statusCode,
          ...metricLabels
        },
        1,
        new Date()
      )
    }

    console.debug('HTTP response', {
      method: resolvedOptions.method,
      url: stripCredentialsFromUrl(response.url),
      status: response.statusCode,
      requestId: options.requestId,
      duration: duration ?? undefined,
      timingPhases: roundProperties(response.timings)
    })

    return response as any
  } catch (err) {
    if (err instanceof HTTPError) {
      const { code, message, response } = err
      // The server responded with a status codes other than 2xx.
      // Check reason.statusCode
      if (trackMetrics) {
        httpRequestCounter.inc(
          {
            statusCode: response.statusCode,
            ...metricLabels
          },
          1,
          new Date()
        )
      }

      const href = response.request && response.url
      console.warn('HTTP StatusCodeError', err, {
        method: options.method,
        statusCode: code,
        requestId: options.requestId,
        url: stripCredentialsFromUrl(href),
        errMsg: message || ''
      })

      throw err
    }

    // The request failed due to technical reasons.
    // reason.cause is the Error object Request would pass into a callback.
    if (trackMetrics) {
      httpRequestCounter.inc(
        {
          statusCode: -1,
          ...metricLabels
        },
        1,
        new Date()
      )
    }

    console.warn('HTTP RequestError', err, {
      method: options.method,
      requestId: options.requestId,
      url: stripCredentialsFromUrl(options.url),
      errMsg: err.message,
      code: err.code,
      errStack: err.stack
    })
    throw err
  }
}

export function withDefaults(optionDefaults: Partial<HttpOptions>): HttpClient {
  return new HttpClient(defaultGot.extend(mapOptions(optionDefaults)))
}

export const httpClient = new HttpClient()
