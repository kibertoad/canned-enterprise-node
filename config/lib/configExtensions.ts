import * as R from 'ramda'
import express from 'express'

import { FeatureToggles } from './configInterface'
import { getFeatures as getBaseFeatures } from './configService'
import { Brand, Locale } from '@tf-pltfrm/core'

const HEADER_OVERWRITE_FEATURES = 'x-overwrite-features'
function extractFeaturesFromHeader(req) {
  const header = req.get(HEADER_OVERWRITE_FEATURES)
  if (!header) return {}
  const featuresWithEqualSignList = splitAndGetPropsWithEqualSign(header)
  const featurePairs = featuresWithEqualSignList.map((prop) => {
    const splitProp = prop.split('=', 2)
    return [splitProp[0], splitProp[1] === 'true']
  })
  return R.fromPairs(featurePairs)
}
const splitAndGetPropsWithEqualSign = (str) => str.split(';').filter((c) => c.indexOf('=') > 0)

export function getFeatures(locale: Locale, brand: Brand, req: express.Request): FeatureToggles {
  const features = getBaseFeatures(locale, brand)
  const featuresToOverwriteFromHeader = extractFeaturesFromHeader(req)
  return (R.mergeDeepRight(features, featuresToOverwriteFromHeader) as unknown) as FeatureToggles
}
