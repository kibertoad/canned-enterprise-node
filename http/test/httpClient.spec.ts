import * as nock from 'nock'
import { NockbackHelper } from 'nockback-harder'
import { httpClient } from '../lib/httpClient'

export function initNockbackHelper(dirname: string, passThroughLocalCall = true): NockbackHelper {
  const helper = new NockbackHelper(nock, dirname + '/__nock-fixtures__', {
    passThroughLocalCall
  })
  return helper
}

describe('httpClient', () => {
  let nockbackHelper: NockbackHelper
  beforeAll(() => {
    nockbackHelper = initNockbackHelper(__dirname)
  })

  describe('GET', () => {
    it('should support GET operation', async () => {
      await nockbackHelper.nockBack('GET_happy.json', async () => {
        const response = await httpClient.get('http://jsonplaceholder.typicode.com/todos/1')
        expect(response.body).toEqual({
          completed: false,
          id: 1,
          title: 'delectus aut autem',
          userId: 1
        })
        expect(response.statusCode).toEqual(200)
        expect(response.url).toEqual('http://jsonplaceholder.typicode.com/todos/1')
      })
    })

    it('should support query string params', async () => {
      await nockbackHelper.nockBack('GET_qs.json', async () => {
        const response = await httpClient.get('http://jsonplaceholder.typicode.com/todos', {
          qs: {
            userId: 1
          }
        })
        expect(response.body).toMatchSnapshot()
        expect(response.statusCode).toEqual(200)
        expect(response.url).toEqual('http://jsonplaceholder.typicode.com/todos?userId=1')
      })
    })

    it('should handle 500 for GET operation', async () => {
      await nockbackHelper.nockBack('GET_500.json', async () => {
        expect.assertions(3)
        try {
          await httpClient.get('http://httpstat.us/500')
        } catch (err) {
          const response = err.response
          expect(response.body).toEqual({
            code: 500,
            description: 'Internal Server Error'
          })
          expect(response.statusMessage).toEqual('Internal Server Error')
          expect(response.statusCode).toEqual(500)
        }
      })
    })

    it('should handle 404 for GET operation', async () => {
      await nockbackHelper.nockBack('GET_404.json', async () => {
        expect.assertions(3)
        try {
          await httpClient.get('http://httpstat.us/404')
        } catch (err) {
          const response = err.response
          expect(response.body).toEqual({
            code: 404,
            description: 'Not Found'
          })
          expect(response.statusMessage).toEqual('Not Found')
          expect(response.statusCode).toEqual(404)
        }
      })
    })
  })

  describe('POST', () => {
    it('should support POST operation', async () => {
      await nockbackHelper.nockBack('POST_happy.json', async () => {
        const response = await httpClient.post('http://jsonplaceholder.typicode.com/todos/', {
          title: 'lorem ipsum'
        })
        expect(response.body).toEqual({
          id: 201,
          title: 'lorem ipsum'
        })
        expect((response as any).req!.method).toEqual('POST')
        expect(response.statusCode).toEqual(201)
        expect(response.url).toEqual('http://jsonplaceholder.typicode.com/todos/')
      })
    })
  })

  describe('PUT', () => {
    it('should support PUT operation', async () => {
      await nockbackHelper.nockBack('PUT_happy.json', async () => {
        const response = await httpClient.put('http://jsonplaceholder.typicode.com/posts/1', {
          id: 1,
          title: 'lorem ipsum'
        })
        expect(response.body).toEqual({
          id: 1,
          title: 'lorem ipsum'
        })
        expect((response as any).req!.method).toEqual('PUT')
        expect(response.statusCode).toEqual(200)
        expect(response.url).toEqual('http://jsonplaceholder.typicode.com/posts/1')
      })
    })
  })

  describe('PATCH', () => {
    it('should support PATCH operation', async () => {
      await nockbackHelper.nockBack('PATCH_happy.json', async () => {
        const response = await httpClient.patch('http://jsonplaceholder.typicode.com/todos/5', {
          id: 5,
          title: 'lorem ipsum'
        })
        expect(response.body).toEqual({
          completed: false,
          userId: 1,
          id: 5,
          title: 'lorem ipsum'
        })
        expect(response.statusCode).toEqual(200)
        expect(response.url).toEqual('http://jsonplaceholder.typicode.com/todos/5')
      })
    })
  })

  describe('DELETE', () => {
    it('should support DELETE operation', async () => {
      await nockbackHelper.nockBack('DELETE_happy.json', async () => {
        const response = await httpClient.delete('http://jsonplaceholder.typicode.com/todos/1')
        expect(response.body).toEqual({})
        expect((response as any).req!.method).toEqual('DELETE')
        expect(response.statusCode).toEqual(200)
        expect(response.url).toEqual('http://jsonplaceholder.typicode.com/todos/1')
      })
    })
  })
})
