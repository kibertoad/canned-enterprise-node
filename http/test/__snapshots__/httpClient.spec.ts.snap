// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`httpClient GET should support query string params 1`] = `
Array [
  Object {
    "completed": false,
    "id": 1,
    "title": "delectus aut autem",
    "userId": 1,
  },
  Object {
    "completed": false,
    "id": 2,
    "title": "quis ut nam facilis et officia qui",
    "userId": 1,
  },
  Object {
    "completed": false,
    "id": 3,
    "title": "fugiat veniam minus",
    "userId": 1,
  },
  Object {
    "completed": true,
    "id": 4,
    "title": "et porro tempora",
    "userId": 1,
  },
  Object {
    "completed": false,
    "id": 5,
    "title": "laboriosam mollitia et enim quasi adipisci quia provident illum",
    "userId": 1,
  },
  Object {
    "completed": false,
    "id": 6,
    "title": "qui ullam ratione quibusdam voluptatem quia omnis",
    "userId": 1,
  },
  Object {
    "completed": false,
    "id": 7,
    "title": "illo expedita consequatur quia in",
    "userId": 1,
  },
  Object {
    "completed": true,
    "id": 8,
    "title": "quo adipisci enim quam ut ab",
    "userId": 1,
  },
  Object {
    "completed": false,
    "id": 9,
    "title": "molestiae perspiciatis ipsa",
    "userId": 1,
  },
  Object {
    "completed": true,
    "id": 10,
    "title": "illo est ratione doloremque quia maiores aut",
    "userId": 1,
  },
  Object {
    "completed": true,
    "id": 11,
    "title": "vero rerum temporibus dolor",
    "userId": 1,
  },
  Object {
    "completed": true,
    "id": 12,
    "title": "ipsa repellendus fugit nisi",
    "userId": 1,
  },
  Object {
    "completed": false,
    "id": 13,
    "title": "et doloremque nulla",
    "userId": 1,
  },
  Object {
    "completed": true,
    "id": 14,
    "title": "repellendus sunt dolores architecto voluptatum",
    "userId": 1,
  },
  Object {
    "completed": true,
    "id": 15,
    "title": "ab voluptatum amet voluptas",
    "userId": 1,
  },
  Object {
    "completed": true,
    "id": 16,
    "title": "accusamus eos facilis sint et aut voluptatem",
    "userId": 1,
  },
  Object {
    "completed": true,
    "id": 17,
    "title": "quo laboriosam deleniti aut qui",
    "userId": 1,
  },
  Object {
    "completed": false,
    "id": 18,
    "title": "dolorum est consequatur ea mollitia in culpa",
    "userId": 1,
  },
  Object {
    "completed": true,
    "id": 19,
    "title": "molestiae ipsa aut voluptatibus pariatur dolor nihil",
    "userId": 1,
  },
  Object {
    "completed": true,
    "id": 20,
    "title": "ullam nobis libero sapiente ad optio sint",
    "userId": 1,
  },
]
`;
